server/handler.go

```diff
@@ -115,26 +153,38 @@ func (h *Handler) ComQuery(
 
 			r = nil
 			proccesedAtLeastOneBatch = true
-
 			continue
 		}
 
-		row, err := rows.Next()
-		if err != nil {
+		select {
+		case err := <-errchan:
 			if err == io.EOF {
-				break
+				break rowLoop
 			}
 
 			return err
-		}
+		case row := <-rowchan:
+			outputRow, err := rowToSQL(schema, row)
+			if err != nil {
+				close(quit)
+				break rowLoop
+			}
 
-		outputRow, err := rowToSQL(schema, row)
-		if err != nil {
-			return err
+			r.Rows = append(r.Rows, outputRow)
+			r.RowsAffected++
+		case <-timer.C:
+			if h.readTimeout != 0 {
+				// Return so Vitess can call the CloseConnection callback
+				err = ErrRowTimeout.New()
+				close(quit)
+				break rowLoop
+			}
 		}
+		timer.Reset(waitTime)
+	}
 
-		r.Rows = append(r.Rows, outputRow)
-		r.RowsAffected++
+	if err != nil {
```

Comment:
where does this err come from? I don't see any err defined in the scope

Reply:
It was for the third case in the select, but should have been removed when I added the `defer`, I've refactored it without it and direct returns on errors to make the select it less jumpy.